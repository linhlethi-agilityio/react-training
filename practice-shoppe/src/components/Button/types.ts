import { BaseVariant } from '@type/variant';
import { ButtonHTMLAttributes, ReactNode } from 'react';

export type Variant = BaseVariant | 'outline';

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant: Variant;
  children: ReactNode;
}
