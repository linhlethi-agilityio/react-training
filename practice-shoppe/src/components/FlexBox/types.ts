import { TUnitValue } from '@type/unitValue';
import { TStyleUnit } from '@helper/transformNumber';

export type TJustifyContent = 'center' | 'space-between' | 'left' | 'right';

export interface IFlexBoxProps {
  gap?: TUnitValue;
  unitFlexBox?: TStyleUnit;
  justifyContent?: TJustifyContent;
  mTop?: TUnitValue;
  mLeft?: TUnitValue;
  mRight?: TUnitValue;
  mBottom?: TUnitValue;
}
