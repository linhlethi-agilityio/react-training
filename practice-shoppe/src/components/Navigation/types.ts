interface NavigationConfig {
  id: number;
  name: string;
  path: string;
}

export interface INavProps {
  className?: string;
  items: NavigationConfig[];
}
