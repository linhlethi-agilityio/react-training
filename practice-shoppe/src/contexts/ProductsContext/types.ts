import { IProduct } from '@type/product';
import { ReactNode } from 'react';

export interface ISortOption {
  _order?: string;
  _sort?: string;
}

export interface IProductsContext {
  state: IProductsState;
  actions: {
    getProductsRequest: () => void;
    getProductsRequestSuccess: (response: IProduct[], totalProducts: number) => void;
    getProductsRequestFail: (errorMessage: string) => void;
    setPage: (pageNumber: number) => void;
    handleSortProducts: (sortOption: ISortOption) => void;
    handleSearchProducts: (searchValue: string) => void;
    resetInitialState: () => void;
  };
}

export interface IContextProviderProps {
  children: ReactNode;
}

export interface IProductsState {
  products: IProduct[];
  totalProducts: number;
  pageNumber: number;
  loading: boolean;
  errorMessage: null | string;
  sortOption: ISortOption;
  searchValue: string;
}
