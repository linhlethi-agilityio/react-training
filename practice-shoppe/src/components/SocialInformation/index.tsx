import { StyledLink } from '@components/Link';
import { memo } from 'react';
import { transformToStyleValue } from '@helper/transformNumber';
import styled from 'styled-components';
import { ISocialInformationProps } from './types';

// https://vitejs.dev/guide/assets.html#new-url-url-import-meta-url
const twitterIcon = new URL('/images/twitterIcon.webp', import.meta.url).href;
const facebookIcon = new URL('/images/facebookIcon.webp', import.meta.url).href;
const instagramIcon = new URL('/images/instagramIcon.webp', import.meta.url).href;
const linkedinIcon = new URL('/images/linkedinIcon.webp', import.meta.url).href;

const SocialInformation = (props: ISocialInformationProps) => {
  const { className, data } = props;

  return (
    <div className={className}>
      <StyledLink to={data.linkedin}>
        <img src={linkedinIcon} alt="linkedinIcon" />
      </StyledLink>
      <StyledLink to={data.facebook}>
        <img src={facebookIcon} alt="facebookIcon" />
      </StyledLink>
      <StyledLink to={data.instagram}>
        <img src={instagramIcon} alt="instagramIcon" />
      </StyledLink>
      <StyledLink to={data.twitter}>
        <img src={twitterIcon} alt="twitterIcon" />
      </StyledLink>
    </div>
  );
};

const StyledSocialInformation = memo(styled(SocialInformation)<ISocialInformationProps>`
  display: flex;
  gap: ${(props) => transformToStyleValue(props.gap)};
`);

export { StyledSocialInformation };
