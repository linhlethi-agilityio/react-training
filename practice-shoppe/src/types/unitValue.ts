import { TWidthValue } from '@helper/transformNumber';

export type TUnitValue = number | TWidthValue;
