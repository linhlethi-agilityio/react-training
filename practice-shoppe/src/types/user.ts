export interface IUser {
  id: number;
  email: string;
  password: string;
}

export interface IUserLogin {
  id: number;
  user: IUser;
  accessToken: string;
}

export type TUserResponse = IUserLogin | null | undefined;
