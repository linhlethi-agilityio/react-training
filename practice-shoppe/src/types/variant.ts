export type BaseVariant = 'primary' | 'secondary';
