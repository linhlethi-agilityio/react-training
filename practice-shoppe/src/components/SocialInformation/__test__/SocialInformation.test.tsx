import { RenderResult, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { StyledSocialInformation } from '@components/SocialInformation';
import { ISocialInformationProps } from '../types';
import { SOCIAL_INFORMATION } from '@constants/baseConfig';

const StyledSocialInformationWithRouter = (props: ISocialInformationProps) => {
  return (
    <BrowserRouter>
      <StyledSocialInformation {...props} />
    </BrowserRouter>
  );
};

describe('testing SocialInformation component', () => {
  let wrapper: RenderResult;

  beforeEach(() => {
    wrapper = render(<StyledSocialInformationWithRouter data={SOCIAL_INFORMATION} />);
  });

  it('matches snapshot', () => {
    const { asFragment } = wrapper;

    expect(asFragment()).toMatchSnapshot();
  });

  it('should be render icon linkedin first', () => {
    const { getAllByRole } = wrapper;

    const links = getAllByRole('link') as HTMLLinkElement[];

    expect(links[0].href === SOCIAL_INFORMATION.linkedin);
  });

  it('should be render icon facebook middle', () => {
    const { getAllByRole } = wrapper;
    const links = getAllByRole('link') as HTMLLinkElement[];

    expect(links[1].href === SOCIAL_INFORMATION.facebook).toBe(true);
  });

  it('should be render icon instagram middle', () => {
    const { getAllByRole } = wrapper;

    const links = getAllByRole('link') as HTMLLinkElement[];

    expect(links[2].href === SOCIAL_INFORMATION.instagram).toBe(true);
  });

  it('should be render icon twitter last', () => {
    const { getAllByRole } = wrapper;

    const links = getAllByRole('link') as HTMLLinkElement[];

    expect(links[3].href === SOCIAL_INFORMATION.twitter).toBe(true);
  });
});
