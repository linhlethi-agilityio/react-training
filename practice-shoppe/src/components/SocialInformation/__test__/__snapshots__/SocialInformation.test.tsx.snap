// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SocialInformation component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dmqHEX caBKCy"
  >
    <a
      class="sc-beqWaB iEMdCi"
      href="https://www.linkedin.com/"
    >
      <img
        alt="linkedinIcon"
        src="http://localhost/images/linkedinIcon.webp"
      />
    </a>
    <a
      class="sc-beqWaB iEMdCi"
      href="https://www.facebook.com/"
    >
      <img
        alt="facebookIcon"
        src="http://localhost/images/facebookIcon.webp"
      />
    </a>
    <a
      class="sc-beqWaB iEMdCi"
      href="https://www.instagram.com/"
    >
      <img
        alt="instagramIcon"
        src="http://localhost/images/instagramIcon.webp"
      />
    </a>
    <a
      class="sc-beqWaB iEMdCi"
      href="https://twitter.com/"
    >
      <img
        alt="twitterIcon"
        src="http://localhost/images/twitterIcon.webp"
      />
    </a>
  </div>
</DocumentFragment>
`;
